src/imars3d/backend/morph/crop.py

```diff
@@ -64,106 +61,150 @@ def crop(
+        expand_ratio,
+        rel_intensity_threshold_air_or_slit,
+        rel_intensity_threshold_fov,
+        rel_intensity_threshold_sample,
+    ) -> np.ndarray:
 
-    Returns
-    -------
-        The crop limits in (left, right, top, bottom) order.
-    """
-    # generate representative image
-    if arrays.ndim == 2:
-        img = arrays
-    elif arrays.ndim == 3:
-        img = arrays.mean(axis=0)
-    else:
-        raise ValueError("Only 2D and 3D arrays are supported.")
-
-    # denoise
-    img = median_filter(img, 9).astype(float)
-    # rescale
-    img = (img - img.min()) / (img.max() - img.min())
-    # estimate background from four stripes near the border
-    left = np.median(img[:, :border_pix])
-    right = np.median(img[:, -border_pix:])
-    top = np.median(img[:border_pix, :])
-    bottom = np.median(img[-border_pix:, :])
-    intensity_bg = np.median([left, right, top, bottom])
-
-    if intensity_bg < rel_intensity_threshold_air_or_slit:
-        # Case 1: slits in, i.e I_inner >> I_outer
-        #     when the background intensity is around the bottom 5% of the
-        #     total dynamic range, we select the pixels with the top 90% intensity
-        #     to be the field of view.
-        ys, xs = np.where(img > rel_intensity_threshold_fov)
-        dx = dy = 0.0  # no expanding needed
-    else:
-        # Case 2: slits out, i.e I_inner < I_outer
-        #     when the estimated background intensity is high (> 5%), we assume
-        #     that the slits are either out or removed from image, therefore the
-        #     detection now is for the object within the FOV.
-        #     since object will absorb neutron, selecting the lower intensity
-        #     region helps us identify the rough bounding box.
-        ys, xs = np.where(img < intensity_bg * rel_intensity_threshold_sample)
-        #
-        width = xs.max() - xs.min()
-        height = ys.max() - ys.min()
-        #
-        dx = width * expand_ratio
-        dy = height * expand_ratio
-
-    # return the limits
-    return (
-        int(max(xs.min() - dx, 0)),
-        int(min(xs.max() + dx, img.shape[1])),
-        int(max(ys.min() - dy, 0)),
-        int(min(ys.max() + dy, img.shape[0])),
-    )
+        if arrays.ndim not in (2, 3):
+            raise ValueError("Only 2D and 3D arrays are supported.")
+
+        # check if crop limits (bounding box) are provided
+        if -1 in crop_limit:
+            crop_limit = crop.detect_bounds(
+                arrays,
+                border_pix,
+                expand_ratio,
+                rel_intensity_threshold_air_or_slit,
+                rel_intensity_threshold_fov,
+                rel_intensity_threshold_sample,
+            )
+        # crop
+        left, right, top, bottom = crop_limit
+        if arrays.ndim == 2:
+            return arrays[top:bottom, left:right]
+        elif arrays.ndim == 3:
+            return arrays[:, top:bottom, left:right]
+
+    def detect_bounds(
```

Comment:
how about making `detect_bounds` a module level function, parallel to `crop` instead of embedded inside as a semi-class method (there is no classmethod decorator here, so it is hard to tell)?